fdbrpc/JsonWebKeySet.cpp

```diff
@@ -0,0 +1,828 @@
+#include <rapidjson/document.h>
+#include <rapidjson/error/en.h>
+#include <rapidjson/writer.h>
+#include <rapidjson/stringbuffer.h>
+#include <array>
+#include <string_view>
+#include <type_traits>
+
+#define JWKS_ERROR(issue, op)                                                                                          \
+	TraceEvent(SevWarnAlways, "JsonWebKeySet" #op "Error").suppressFor(10).detail("Issue", issue)
+#define JWKS_PARSE_ERROR(issue) JWKS_ERROR(issue, Parse)
+#define JWKS_WRITE_ERROR(issue) JWKS_ERROR(issue, Write)
+
+#define JWK_PARSE_ERROR(issue)                                                                                         \
+	TraceEvent(SevWarnAlways, "JsonWebKeyParseError")                                                                  \
+	    .suppressFor(10)                                                                                               \
+	    .detail("Issue", issue)                                                                                        \
+	    .detail("KeyIndexBase0", keyIndex)
+#define JWK_WRITE_ERROR(issue)                                                                                         \
+	TraceEvent(SevWarnAlways, "JsonWebKeyWriteError")                                                                  \
+	    .suppressFor(10)                                                                                               \
+	    .detail("Issue", issue)                                                                                        \
+	    .detail("KeyName", keyName.toString())
+#define JWK_ERROR_OSSL(issue, op)                                                                                      \
+	do {                                                                                                               \
+		char buf[256]{                                                                                                 \
+			0,                                                                                                         \
+		};                                                                                                             \
+		if (auto err = ::ERR_get_error()) {                                                                            \
+			::ERR_error_string_n(err, buf, sizeof(buf));                                                               \
+		}                                                                                                              \
+		JWK_##op##_ERROR(issue).detail("OpenSSLError", static_cast<char const*>(buf));                                 \
+	} while (0)
+#define JWK_PARSE_ERROR_OSSL(issue) JWK_ERROR_OSSL(issue, PARSE)
+#define JWK_WRITE_ERROR_OSSL(issue) JWK_ERROR_OSSL(issue, WRITE)
+
+// introduces a new variable [member] to enclosing scope, wrapped in Optional if optional == true
+#define DECL_JWK_STR_MEMBER(valueObj, member, optional)                                                                \
```

Comment:
Seems like a lot of code for a macro. Can we extract a reusable function/template here?
Also check DECL_DECODED_BN_MEMBER_OPTIONAL and DECL_DECODED_BN_MEMBER_REQUIRED

Reply:
Done!